machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,207 @@
+	// +optional
+	BaseImagePullSecret *ImageSecretObjectReference `json:"baseImagePullSecret,omitempty"`
+	// renderedImagePushSecret is the secret used to connect to a user registry.
+	// The final image push and pull secrets should be separate and assume the principal of least privilege.
+	// The push secret with write privilege is only required to be present on the node hosting the MachineConfigController pod.
+	// The pull secret with read only privileges is required on all nodes.
+	// By separating the two secrets, the risk of write credentials becoming compromised is reduced.
+	// +required
+	RenderedImagePushSecret ImageSecretObjectReference `json:"renderedImagePushSecret"`
+	// renderedImagePushSpec describes the location of the final image.
+	// The MachineOSConfig object will use the in cluster image registry configuration.
+	// If you wish to use a mirror or any other settings specific to registries.conf, please specify those in the cluster wide registries.conf via the cluster image.config, ImageContentSourcePolicies, ImageDigestMirrorSet, or ImageTagMirrorSet objects.
+	// The format of the image push spec is: host[:port][/namespace]/name:<tag> or svc_name.namespace.svc[:port]/repository/name:<tag>.
+	// The length of the push spec must be between 1 to 447 characters.
+	// +required
+	RenderedImagePushSpec ImageTagFormat `json:"renderedImagePushSpec"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// This is also commonly called a Dockerfile and you can treat it as such. The content is the content of your Dockerfile.
+	// See https://github.com/containers/common/blob/main/docs/Containerfile.5.md for the spec reference.
+	// This is a list indexed by architecture name (e.g. AMD64), and allows specifying one containerFile per arch, up to 4.
+	// +patchMergeKey=containerfileArch
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=containerfileArch
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=4
+	// +optional
+	Containerfile []MachineOSContainerfile `json:"containerFile" patchStrategy:"merge" patchMergeKey:"containerfileArch"`
+}
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
+	// observedGeneration represents the generation of the MachineOSConfig object observed by the Machine Config Operator's build controller.
+	// +kubebuilder:validation:XValidation:rule="self >= oldSelf", message="observedGeneration must not move backwards"
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// currentImagePullSpec is the fully qualified image pull spec used by the MCO to pull down the new OSImage. This includes the sha256 image digest.
```

Comment:
Is this updated over time, if the MOSC triggers new builds? Can we explain when this field might change, once populated?

Reply:
This is part of the status of the MOSC, I would expect this to always point to the latest successfully built MOSB, right?

Reply:
the controller does set it to the MOSB's "push spec": `mosc.Status.CurrentImagePullspec = mosb.Status.FinalImagePushspec`

which I think is populated when a build is successful. So this can be considered tracking the corresponding MOSB with a successfully created push spec